daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/DelimiterUnparsers.scala

```diff
@@ -56,7 +56,7 @@ class DelimiterTextUnparser(override val context: TermRuntimeData, delimiterType
       else localDelimNode.terminator
     }
 
-    if (delimDFAs.isEmpty) Assert.invariantFailed("Expected a delimiter of type " + delimiterType + " on the stack, but was not found.")
+    if (delimDFAs.length == 0) Assert.invariantFailed("Expected a delimiter of type " + delimiterType + " on the stack, but was not found.")
```

Comment:
Is this stylistic, or is there actually an advantage to using length vs. isEmpty?

Reply:
Yep, there's an actual advantage here. delimDFAs is type ``Array[T]``. Scala will normally do the right than and represent this in byte code as a java array[]. All is good. However, if you try to use any functions that are normally Scala Sequence-y type things (e.g. isEmpty, .size), Scala will implicitly wrap the array in an ``ArrayOps``. This gives you all the nice scala-y stuff that were used to. However, this means we are allocating an ArrayOps just do figure out if the array is empty or not. While it's convenient and looks nice, it added does add allocations. I'm not sure if it really affected the performance, but this was called a ton, and made the profiling very noise, so this at least helps to get rid of some of that noise.

Reply:
Deserves a note in our performance page on the wiki saying when to prefer java arrays to sequences, and then ifyou are using java arrays, how to use them so as to avoid the downsides of this particular flavor of auto-conversion.

Reply:
Agreed, I'll add a write to our wiki page about all the tricks I'm finding to reduce allocations.

Reply:
Tips I found are now on the wiki on our [Coding for Performance](https://cwiki.apache.org/confluence/display/DAFFODIL/Coding+for+Performance) page.